File: c/grains/grains.c
#include "grains.h"

uint64_t square(uint8_t index) {
    if (index < 1 || index > 64) return 0;
    
    uint64_t accum = 1;
    for (uint8_t i=1; i<index; i++) accum *= 2;
    return accum;
}

uint64_t total(void) {
    return 0xffffffffffffffff;
}

File: julia/grains/grains.jl
"""Calculate the number of grains on square `square`."""
function on_square(square)
    if !(1 ≤ square ≤ 64)
        throw(DomainError(square, "Invalid square"))
    end
    
    big(2) ^ (square - 1)
end

"""Calculate the total number of grains after square `square`."""
function total_after(square)
    if !(1 ≤ square ≤ 64)
        throw(DomainError(square, "Invalid square"))
    end
    
    sum(on_square(n) for n in 1:square)
end
